src/main/kotlin/org/rust/lang/core/psi/util/RustAttrExtensions.kt

```diff
@@ -0,0 +1,65 @@
+package org.rust.lang.core.psi.util
+
+import org.rust.lang.core.psi.*
+
+/**
+ * Appends a new meta item `new` to an attribute with a list of arguments:
+ *
+ * ```
+ * #[attr]              ->  #[attr(new)]
+ * #[attr="something"]  ->  #[attr(new)]
+ * #[attr(something)]   ->  #[attr(something, new)]
+ * #[attr(something]    ->  #[attr(something, new]
+ * #[attr()]            ->  #[attr(new)]
+ * #[attr(]             ->  #[attr(new]
+ * ```
+ *
+ * @return `meta` argument or its copy if insertion was successful, null otherwise.
+ */
+fun RustAttr.appendMetaToList(meta: RustMetaItem): RustMetaItem? {
+    // extract attribute internals
+    val attrMeta = when (this) {
+        is RustInnerAttr -> metaItem
+        is RustOuterAttr -> metaItem
+        else             -> return null
+    } ?: return null
+
+    if (attrMeta.lparen == null && attrMeta.rparen == null) {
+        // if both parens are missing, remove everything inside the attribute except its identifier and add `meta`
+        // between parens to it
+        if (attrMeta.identifier.nextSibling != null) {
+            attrMeta.deleteChildRange(attrMeta.identifier.nextSibling, attrMeta.lastChild)
+        }
+
+        val tempMeta = RustElementFactory.createMeta(this.project, "whatever(item)") ?: return null
+        tempMeta.metaItem!!.replace(meta)
+
+        val newLeftParen = attrMeta.addRangeAfter(tempMeta.lparen, tempMeta.rparen, attrMeta.identifier)
+        return newLeftParen.nextSibling as RustMetaItem
+
+    } else if (attrMeta.lparen != null) {
+        // if left paren is present, add `meta` to the end of the list of items
+        val leftParen = attrMeta.lparen!!
+
+        if (leftParen.nextSibling != null && leftParen.nextSibling != attrMeta.rparen) {
+            // if there are other meta items in parens, add `meta` after the last one of them
+            val elementToInsertAfter = if (attrMeta.rparen != null)
+                attrMeta.rparen!!.prevSibling!!
+            else
+                attrMeta.lastChild!!
+
+            // we also need a comma, so first we insert a comma and a temporary meta item from another well-formed meta
+            val tempMeta = RustElementFactory.createMeta(this.project, "whatever(item,item)") ?: return null
+            val start = tempMeta.lparen!!.nextSibling!!.nextSibling!!
```

Comment:
When I was writing the initial implementation, something like `derive()` wasn't a valid `RustMetaItem` :) But since #319 is closed I guess it could work.

And no, I didn't handle adding a comma :( I was under impression that it was forbidden by the grammar as well but it seems it is not so. Yes, I guess your approach does look better and more robust.

Actually, I was thinking of adding an intention like "Add derive clauses for common traits" which should add well-known traits like `Eq`, `PartialEq` and such to the `derive` clause, possibly presenting a list to the user to select the necessary ones and/or deducing possible ones automatically. But this, I guess, would require more thorough analysis of struct contents, like which traits inner types implement. Anyway, this is a matter for another pull request, I believe :) Just wanted to share an idea.